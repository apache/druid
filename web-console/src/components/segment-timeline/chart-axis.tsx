/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { Axis } from 'd3-axis';
import { select } from 'd3-selection';

interface ChartAxisProps {
  className?: string;
  transform?: string;
  axis: Axis<any>;
}

export const ChartAxis = function ChartAxis(props: ChartAxisProps) {
  const { transform, axis, className } = props;
  return (
    <g
      className={`chart-axis ${className}`}
      transform={transform}
      ref={node => select(node).call(axis as any)}
    />
  );
};
