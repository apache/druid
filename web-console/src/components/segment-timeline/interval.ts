/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { Duration } from 'chronoshift';
import { sum } from 'd3-array';

import { formatBytes, formatInteger } from '../../utils';

export type IntervalStat = 'segments' | 'size' | 'rows';

export const INTERVAL_STATS: IntervalStat[] = ['segments', 'size', 'rows'];

export function getIntervalStatTitle(intervalStat: IntervalStat): string {
  switch (intervalStat) {
    case 'segments':
      return 'Num. segments';

    case 'size':
      return 'Size';

    case 'rows':
      return 'Rows';

    default:
      return intervalStat;
  }
}

export function aggregateSegmentStats(
  xs: readonly Record<IntervalStat, number>[],
): Record<IntervalStat, number> {
  return {
    segments: sum(xs, s => s.segments),
    size: sum(xs, s => s.size),
    rows: sum(xs, s => s.rows),
  };
}

export function formatIntervalStat(stat: IntervalStat, n: number) {
  switch (stat) {
    case 'segments':
    case 'rows':
      return formatInteger(n);

    case 'size':
      return formatBytes(n);

    default:
      return '';
  }
}

export interface IntervalRow extends Record<IntervalStat, number> {
  start: Date;
  end: Date;
  datasource: string;
  realtime: boolean;
  originalTimeSpan: Duration;
}

export interface TrimmedIntervalRow extends IntervalRow {
  shownDays: number;
  normalized: Record<IntervalStat, number>;
}

export interface IntervalBar extends TrimmedIntervalRow {
  offset: Record<IntervalStat, number>;
}

export function formatIsoDateOnly(date: Date): string {
  return date.toISOString().slice(0, 10);
}
