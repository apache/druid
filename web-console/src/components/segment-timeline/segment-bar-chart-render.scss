/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

@import '../../variables';

.segment-bar-chart-render {
  position: relative;
  overflow: hidden;

  @keyframes pulseOpacity {
    0% {
      opacity: 0.8;
    }
    100% {
      opacity: 0.95;
    }
  }

  svg {
    position: absolute;

    text {
      user-select: none;
    }

    .hover-highlight {
      fill: white;
      fill-opacity: 0.1;
    }

    .hovered-bar {
      fill: none;
      stroke: #ffffff;
      stroke-width: 1.5px;
    }

    .selection {
      fill: transparent;
      stroke: #ffffff;
      stroke-width: 1px;
      opacity: 0.8;

      &.finalized {
        opacity: 1;
      }
    }

    .shifter {
      fill: white;
      fill-opacity: 0.2;
      filter: blur(1px);
    }

    .time-shift-indicator {
      fill: white;
      fill-opacity: 0.001;
      cursor: grab;

      &:hover {
        fill-opacity: 0.1;
      }

      &.shifting {
        fill-opacity: 0.2;
        cursor: grabbing;
      }
    }

    .h-gridline {
      line {
        stroke: $white;
        stroke-dasharray: 5, 5;
        opacity: 0.5;
      }
    }

    .now-line {
      stroke: $orange4;
      stroke-dasharray: 2, 2;
      opacity: 0.7;
    }

    .bar-unit {
      &.realtime {
        animation: pulseOpacity 3s alternate infinite;
      }
    }
  }

  .rule-tape {
    position: absolute;
    top: 5px;
    height: 15px;
    font-size: 10px;

    .rule-error {
      @include pin-full();
      background-color: $red3;
      color: $white;
    }

    .load-rule {
      position: absolute;
      overflow: hidden;
      padding-left: 2px;
      border-left: 1px solid $dark-gray2;
      border-right: 1px solid $dark-gray2;
      top: 0;
      height: 100%;
      text-overflow: ellipsis;

      &.load {
        background-color: $green1;

        &:nth-child(even) {
          background-color: $green3;
        }
      }

      &.drop {
        background-color: $dark-gray5;

        &:nth-child(even) {
          background-color: $gray1;
        }
      }

      &.broadcast {
        background-color: $indigo1;

        &:nth-child(even) {
          background-color: $indigo3;
        }
      }
    }
  }

  .empty-placeholder {
    @include pin-full;
    display: flex;
    align-items: center;
    justify-content: center;
    font-size: 20px;
    user-select: none;
    pointer-events: none;
  }
}

.segment-bar-chart-bubble {
  .button-bar {
    padding-top: 5px;
    display: flex;
    gap: 5px;
  }
}
