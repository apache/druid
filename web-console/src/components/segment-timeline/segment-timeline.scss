/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

.segment-timeline {
  display: grid;
  grid-template-columns: 85% 15%;
  height: 100%;

  .loader {
    width: 85%;
  }

  .loading-error {
    position: fixed;
    top: 50%;
    left: 50%;
    -webkit-transform: translate(-50%, -50%);
    transform: translate(-50%, -50%);
  }

  .bar-chart-tooltip {
    margin-left: 53px;

    div {
      display: inline-block;
      width: 230px;
    }
  }

  .no-data-text {
    position: absolute;
    left: 30vw;
    top: 15vh;
    font-size: 20px;
  }

  .side-control {
    padding-left: 1vw;
    padding-top: 5vh;
  }
}
