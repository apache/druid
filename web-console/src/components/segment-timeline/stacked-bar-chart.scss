/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

.stacked-bar-chart {
  position: relative;
  overflow: hidden;

  .bar-chart-tooltip {
    position: absolute;
    left: 100px;
    right: 0;

    div {
      display: inline-block;
      width: 230px;
    }
  }

  svg {
    position: absolute;

    .hovered-bar {
      fill: transparent;
      stroke: #ffffff;
      stroke-width: 1.5px;
    }

    .gridline-x {
      line {
        stroke-dasharray: 5, 5;
        opacity: 0.5;
      }
    }
  }
}
