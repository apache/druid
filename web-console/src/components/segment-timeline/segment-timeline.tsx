/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { FormGroup, HTMLSelect, Radio, RadioGroup } from '@blueprintjs/core';
import { AxisScale } from 'd3-axis';
import { scaleLinear, scaleTime } from 'd3-scale';
import React from 'react';

import { Api } from '../../singletons';
import { Capabilities } from '../../utils';
import { formatBytes, queryDruidSql, QueryManager, uniq } from '../../utils/index';
import { StackedBarChart } from '../../visualization/stacked-bar-chart';
import { Loader } from '../loader/loader';

import './segment-timeline.scss';

interface SegmentTimelineProps {
  capabilities: Capabilities;
  chartHeight: number;
  chartWidth: number;

  // For testing:
  dataQueryManager?: QueryManager<{ capabilities: Capabilities; timeSpan: number }, any>;
}

type ActiveDataType = 'sizeData' | 'countData';

interface SegmentTimelineState {
  data?: Record<string, any>;
  datasources: string[];
  stackedData?: Record<string, BarUnitData[]>;
  singleDatasourceData?: Record<string, Record<string, BarUnitData[]>>;
  activeDatasource: string | null;
  activeDataType: ActiveDataType;
  dataToRender: BarUnitData[];
  timeSpan: number; // by months
  loading: boolean;
  error?: Error;
  xScale: AxisScale<Date> | null;
  yScale: AxisScale<number> | null;
  dStart: Date;
  dEnd: Date;
}

interface BarChartScales {
  xScale: AxisScale<Date>;
  yScale: AxisScale<number>;
}

export interface BarUnitData {
  x: number;
  y: number;
  y0?: number;
  width: number;
  datasource: string;
  color: string;
}

export interface BarChartMargin {
  top: number;
  right: number;
  bottom: number;
  left: number;
}

interface IntervalRow {
  start: string;
  end: string;
  datasource: string;
  count: number;
  size: number;
}

const DEFAULT_TIME_SPAN_MONTHS = 3;

export class SegmentTimeline extends React.PureComponent<
  SegmentTimelineProps,
  SegmentTimelineState
> {
  static COLORS = [
    '#b33040',
    '#d25c4d',
    '#f2b447',
    '#d9d574',
    '#4FAA7E',
    '#57ceff',
    '#789113',
    '#098777',
    '#b33040',
    '#d2757b',
    '#f29063',
    '#d9a241',
    '#80aa61',
    '#c4ff9e',
    '#915412',
    '#87606c',
  ];

  static getColor(index: number): string {
    return SegmentTimeline.COLORS[index % SegmentTimeline.COLORS.length];
  }

  static processRawData(data: IntervalRow[]) {
    if (data === null) return [];

    const countData: Record<string, any> = {};
    const sizeData: Record<string, any> = {};
    data.forEach(entry => {
      const start = entry.start;
      const day = start.split('T')[0];
      const datasource = entry.datasource;
      const count = entry.count;
      const segmentSize = entry.size;
      if (countData[day] === undefined) {
        countData[day] = {
          day,
          [datasource]: count,
          total: count,
        };
        sizeData[day] = {
          day,
          [datasource]: segmentSize,
          total: segmentSize,
        };
      } else {
        const countDataEntry = countData[day][datasource];
        countData[day][datasource] = count + (countDataEntry === undefined ? 0 : countDataEntry);
        const sizeDataEntry = sizeData[day][datasource];
        sizeData[day][datasource] = segmentSize + (sizeDataEntry === undefined ? 0 : sizeDataEntry);
        countData[day].total += count;
        sizeData[day].total += segmentSize;
      }
    });

    const countDataArray = Object.keys(countData)
      .reverse()
      .map((time: any) => {
        return countData[time];
      });

    const sizeDataArray = Object.keys(sizeData)
      .reverse()
      .map((time: any) => {
        return sizeData[time];
      });

    return { countData: countDataArray, sizeData: sizeDataArray };
  }

  static calculateStackedData(
    data: Record<string, any>,
    datasources: string[],
  ): Record<string, BarUnitData[]> {
    const newStackedData: Record<string, BarUnitData[]> = {};
    Object.keys(data).forEach((type: any) => {
      const stackedData: any = data[type].map((d: any) => {
        let y0 = 0;
        return datasources.map((datasource: string, i) => {
          const barUnitData = {
            x: d.day,
            y: d[datasource] === undefined ? 0 : d[datasource],
            y0,
            datasource,
            color: SegmentTimeline.getColor(i),
          };
          y0 += d[datasource] === undefined ? 0 : d[datasource];
          return barUnitData;
        });
      });
      newStackedData[type] = stackedData.flat();
    });

    return newStackedData;
  }

  static calculateSingleDatasourceData(
    data: Record<string, any>,
    datasources: string[],
  ): Record<string, Record<string, BarUnitData[]>> {
    const singleDatasourceData: Record<string, Record<string, BarUnitData[]>> = {};
    Object.keys(data).forEach(dataType => {
      singleDatasourceData[dataType] = {};
      datasources.forEach((datasource, i) => {
        const currentData = data[dataType];
        if (currentData.length === 0) return;
        const dataResult = currentData.map((d: any) => {
          let y = 0;
          if (d[datasource] !== undefined) {
            y = d[datasource];
          }
          return {
            x: d.day,
            y,
            datasource,
            color: SegmentTimeline.getColor(i),
          };
        });
        if (!dataResult.every((d: any) => d.y === 0)) {
          singleDatasourceData[dataType][datasource] = dataResult;
        }
      });
    });

    return singleDatasourceData;
  }

  private dataQueryManager: QueryManager<{ capabilities: Capabilities; timeSpan: number }, any>;
  private chartMargin = { top: 20, right: 10, bottom: 20, left: 10 };

  constructor(props: SegmentTimelineProps) {
    super(props);
    const dStart = new Date();
    const dEnd = new Date();
    dStart.setMonth(dStart.getMonth() - DEFAULT_TIME_SPAN_MONTHS);
    this.state = {
      data: {},
      datasources: [],
      stackedData: {},
      singleDatasourceData: {},
      dataToRender: [],
      activeDatasource: null,
      activeDataType: 'sizeData',
      timeSpan: DEFAULT_TIME_SPAN_MONTHS,
      loading: true,
      xScale: null,
      yScale: null,
      dEnd: dEnd,
      dStart: dStart,
    };

    this.dataQueryManager =
      props.dataQueryManager ||
      new QueryManager({
        processQuery: async ({ capabilities, timeSpan }) => {
          let intervals: IntervalRow[];
          let datasources: string[];
          if (capabilities.hasSql()) {
            const query = `
SELECT
  "start", "end", "datasource",
  COUNT(*) AS "count", SUM("size") as "size"
FROM sys.segments
WHERE "start" > TIME_FORMAT(TIMESTAMPADD(MONTH, -${timeSpan}, CURRENT_TIMESTAMP), 'yyyy-MM-dd''T''hh:mm:ss.SSS')
GROUP BY 1, 2, 3
ORDER BY "start" DESC`;

            intervals = await queryDruidSql({ query });
            datasources = uniq(intervals.map(r => r.datasource));
          } else if (capabilities.hasCoordinatorAccess()) {
            const before = new Date();
            before.setMonth(before.getMonth() - timeSpan);
            const beforeIso = before.toISOString();

            datasources = (await Api.instance.get(`/druid/coordinator/v1/datasources`)).data;
            intervals = (await Promise.all(
              datasources.map(async datasource => {
                const intervalMap = (await Api.instance.get(
                  `/druid/coordinator/v1/datasources/${datasource}/intervals?simple`,
                )).data;

                return Object.keys(intervalMap)
                  .map(interval => {
                    const [start, end] = interval.split('/');
                    const { count, size } = intervalMap[interval];
                    return {
                      start,
                      end,
                      datasource,
                      count,
                      size,
                    };
                  })
                  .filter(a => beforeIso < a.start);
              }),
            ))
              .flat()
              .sort((a, b) => b.start.localeCompare(a.start));
          } else {
            throw new Error(`must have SQL or coordinator access`);
          }

          const data = SegmentTimeline.processRawData(intervals);
          const stackedData = SegmentTimeline.calculateStackedData(data, datasources);
          const singleDatasourceData = SegmentTimeline.calculateSingleDatasourceData(
            data,
            datasources,
          );
          return { data, datasources, stackedData, singleDatasourceData };
        },
        onStateChange: ({ data, loading, error }) => {
          this.setState({
            data: data ? data.data : undefined,
            datasources: data ? data.datasources : [],
            stackedData: data ? data.stackedData : undefined,
            singleDatasourceData: data ? data.singleDatasourceData : undefined,
            loading,
            error,
          });
        },
      });
  }

  componentDidMount(): void {
    const { capabilities } = this.props;
    const { timeSpan } = this.state;

    this.dataQueryManager.runQuery({ capabilities, timeSpan });
  }

  componentWillUnmount(): void {
    this.dataQueryManager.terminate();
  }

  componentDidUpdate(prevProps: SegmentTimelineProps, prevState: SegmentTimelineState): void {
    const { activeDatasource, activeDataType, singleDatasourceData, stackedData } = this.state;
    if (
      prevState.data !== this.state.data ||
      prevState.activeDataType !== this.state.activeDataType ||
      prevState.activeDatasource !== this.state.activeDatasource ||
      prevProps.chartWidth !== this.props.chartWidth ||
      prevProps.chartHeight !== this.props.chartHeight
    ) {
      const scales: BarChartScales | undefined = this.calculateScales();
      let dataToRender: BarUnitData[] | undefined;
      dataToRender = activeDatasource
        ? singleDatasourceData
          ? singleDatasourceData[activeDataType][activeDatasource]
          : undefined
        : stackedData
        ? stackedData[activeDataType]
        : undefined;

      if (scales && dataToRender) {
        this.setState({
          dataToRender,
          xScale: scales.xScale,
          yScale: scales.yScale,
        });
      }
    }
  }

  private calculateScales(): BarChartScales | undefined {
    const { chartWidth, chartHeight } = this.props;
    const {
      data,
      activeDataType,
      activeDatasource,
      singleDatasourceData,
      dStart,
      dEnd,
    } = this.state;
    if (!data || !Object.keys(data).length) return;
    const activeData = data[activeDataType];
    const xDomain: Date[] = [dStart, dEnd];
    let yDomain: number[] = [
      0,
      activeData.length === 0
        ? 0
        : activeData.reduce((max: any, d: any) => (max.total > d.total ? max : d)).total,
    ];

    if (
      activeDatasource !== null &&
      singleDatasourceData![activeDataType][activeDatasource] !== undefined
    ) {
      yDomain = [
        0,
        singleDatasourceData![activeDataType][activeDatasource].reduce((max: any, d: any) =>
          max.y > d.y ? max : d,
        ).y,
      ];
    }

    const xScale: AxisScale<Date> = scaleTime()
      .domain(xDomain)
      .range([0, chartWidth - this.chartMargin.left - this.chartMargin.right]);

    const yScale: AxisScale<number> = scaleLinear()
      .rangeRound([chartHeight - this.chartMargin.top - this.chartMargin.bottom, 0])
      .domain(yDomain);

    return {
      xScale,
      yScale,
    };
  }

  onTimeSpanChange = (e: any) => {
    const dStart = new Date();
    const dEnd = new Date();
    const capabilities = this.props.capabilities;
    const timeSpan = parseInt(e, 10) || DEFAULT_TIME_SPAN_MONTHS;
    dStart.setMonth(dStart.getMonth() - timeSpan);
    this.setState({
      timeSpan: e,
      loading: true,
      dStart,
      dEnd,
    });
    this.dataQueryManager.runQuery({ capabilities, timeSpan });
  };

  formatTick = (n: number) => {
    const { activeDataType } = this.state;
    if (activeDataType === 'countData') {
      return n.toString();
    } else {
      return formatBytes(n);
    }
  };

  renderStackedBarChart() {
    const { chartWidth, chartHeight } = this.props;
    const {
      loading,
      dataToRender,
      activeDataType,
      error,
      xScale,
      yScale,
      data,
      activeDatasource,
      dStart,
      dEnd,
    } = this.state;
    if (loading) {
      return (
        <div>
          <Loader loading={loading} />
        </div>
      );
    }

    if (error) {
      return (
        <div>
          <span className={'no-data-text'}>Error when loading data: {error.message}</span>
        </div>
      );
    }

    if (xScale === null || yScale === null) {
      return (
        <div>
          <span className={'no-data-text'}>Error when calculating scales</span>
        </div>
      );
    }

    if (data![activeDataType].length === 0) {
      return (
        <div>
          <span className={'no-data-text'}>No data available for the time span selected</span>
        </div>
      );
    }

    if (
      activeDatasource !== null &&
      data![activeDataType].every((d: any) => d[activeDatasource] === undefined)
    ) {
      return (
        <div>
          <span className={'no-data-text'}>
            No data available for <i>{activeDatasource}</i>
          </span>
        </div>
      );
    }

    const millisecondsPerDay = 24 * 60 * 60 * 1000;
    const barCounts = (dEnd.getTime() - dStart.getTime()) / millisecondsPerDay;
    const barWidth = (chartWidth - this.chartMargin.left - this.chartMargin.right) / barCounts;
    return (
      <StackedBarChart
        dataToRender={dataToRender}
        svgHeight={chartHeight}
        svgWidth={chartWidth}
        margin={this.chartMargin}
        changeActiveDatasource={(datasource: string) =>
          this.setState(prevState => ({
            activeDatasource: prevState.activeDatasource ? null : datasource,
          }))
        }
        activeDataType={activeDataType}
        formatTick={(n: number) => this.formatTick(n)}
        xScale={xScale}
        yScale={yScale}
        barWidth={barWidth}
      />
    );
  }

  render(): JSX.Element {
    const { datasources, activeDataType, activeDatasource, timeSpan } = this.state;

    return (
      <div className={'segment-timeline app-view'}>
        {this.renderStackedBarChart()}
        <div className={'side-control'}>
          <FormGroup>
            <RadioGroup
              onChange={(e: any) => this.setState({ activeDataType: e.target.value })}
              selectedValue={activeDataType}
            >
              <Radio label={'Total size'} value={'sizeData'} />
              <Radio label={'Segment count'} value={'countData'} />
            </RadioGroup>
          </FormGroup>

          <FormGroup label={'Datasource:'}>
            <HTMLSelect
              onChange={(e: any) =>
                this.setState({
                  activeDatasource: e.target.value === 'all' ? null : e.target.value,
                })
              }
              value={activeDatasource == null ? 'all' : activeDatasource}
              fill
            >
              <option value={'all'}>Show all</option>
              {datasources.map(d => {
                return (
                  <option key={d} value={d}>
                    {d}
                  </option>
                );
              })}
            </HTMLSelect>
          </FormGroup>

          <FormGroup label={'Period:'}>
            <HTMLSelect
              onChange={(e: any) => this.onTimeSpanChange(e.target.value)}
              value={timeSpan}
              fill
            >
              <option value={1}>1 months</option>
              <option value={3}>3 months</option>
              <option value={6}>6 months</option>
              <option value={9}>9 months</option>
              <option value={12}>1 year</option>
            </HTMLSelect>
          </FormGroup>
        </div>
      </div>
    );
  }
}
