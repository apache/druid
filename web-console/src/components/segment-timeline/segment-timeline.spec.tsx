/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { render } from '@testing-library/react';
import { mount } from 'enzyme';
import React from 'react';

import { QueryManager } from '../../utils';
import { Capabilities } from '../../utils';

import { SegmentTimeline } from './segment-timeline';

describe('Segment Timeline', () => {
  it('matches snapshot', () => {
    const segmentTimeline = (
      <SegmentTimeline capabilities={Capabilities.FULL} chartHeight={100} chartWidth={100} />
    );
    const { container } = render(segmentTimeline);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('queries 3 months of data by default', () => {
    const dataQueryManager = new MockDataQueryManager();
    const segmentTimeline = (
      <SegmentTimeline
        capabilities={Capabilities.FULL}
        chartHeight={100}
        chartWidth={100}
        dataQueryManager={dataQueryManager}
      />
    );
    render(segmentTimeline);

    // Ideally, the test should verify the rendered bar graph to see if the bars
    // cover the selected period. Since the unit test does not have a druid
    // instance to query from, just verify the query has the correct time span.
    expect(dataQueryManager.queryTimeSpan).toBe(3);
  });

  it('queries matching time span when new period is selected from dropdown', () => {
    const dataQueryManager = new MockDataQueryManager();
    const segmentTimeline = (
      <SegmentTimeline
        capabilities={Capabilities.FULL}
        chartHeight={100}
        chartWidth={100}
        dataQueryManager={dataQueryManager}
      />
    );
    const wrapper = mount(segmentTimeline);
    const selects = wrapper.find('select');
    expect(selects.length).toBe(2); // Datasource & Period
    const periodSelect = selects.at(1);
    const newTimeSpanMonths = 6;
    periodSelect.simulate('change', { target: { value: newTimeSpanMonths } });

    // Ideally, the test should verify the rendered bar graph to see if the bars
    // cover the selected period. Since the unit test does not have a druid
    // instance to query from, just verify the query has the correct time span.
    expect(dataQueryManager.queryTimeSpan).toBe(newTimeSpanMonths);
  });
});

/**
 * Mock the data query manager, since the unit test does not have a druid instance
 */
class MockDataQueryManager extends QueryManager<
  { capabilities: Capabilities; timeSpan: number },
  any
> {
  queryTimeSpan?: number;

  constructor() {
    super({
      processQuery: async ({ timeSpan }) => {
        this.queryTimeSpan = timeSpan;
      },
      debounceIdle: 0,
      debounceLoading: 0,
    });
  }
}
