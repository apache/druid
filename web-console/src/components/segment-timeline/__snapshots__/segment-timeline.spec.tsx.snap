// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Segment Timeline matches snapshot 1`] = `
<div
  class="segment-timeline app-view"
>
  <div>
    <div
      class="loader"
    >
      <div
        class="loader-logo"
      >
        <svg
          viewBox="0 0 100 100"
        >
          <path
            class="one"
            d="M54.2,69.8h-2.7c-0.7,0-1.3-0.6-1.3-1.3c0-0.7,0.6-1.3,1.3-1.3h2.7c11.5,0,23.8-7.4,23.8-23.7
          c0-9.1-6.9-15.8-16.4-15.8H38c-0.7,0-1.3-0.6-1.3-1.3s0.6-1.3,1.3-1.3h23.6c5.3,0,10.1,1.9,13.6,5.3c3.5,3.4,5.4,8,5.4,13.1
          c0,6.6-2.3,13-6.3,17.7C69.5,66.8,62.5,69.8,54.2,69.8z"
          />
          <path
            class="two"
            d="M55.7,59.5h-26c-0.7,0-1.3-0.6-1.3-1.3c0-0.7,0.6-1.3,1.3-1.3h26c7.5,0,11.5-5.8,11.5-11.5
            c0-4.2-3.2-7.3-7.7-7.3h-26c-0.7,0-1.3-0.6-1.3-1.3s0.6-1.3,1.3-1.3h26c5.9,0,10.3,4.3,10.3,9.9c0,3.7-1.3,7.2-3.7,9.8
            C63.5,58,59.9,59.5,55.7,59.5z"
          />
          <path
            class="three"
            d="M27.2,38h-6.3c-0.7,0-1.3-0.6-1.3-1.3s0.6-1.3,1.3-1.3h6.3c0.7,0,1.3,0.6,1.3,1.3S27.9,38,27.2,38z"
          />
          <path
            class="four"
            d="M45.1,69.8h-5.8c-0.7,0-1.3-0.6-1.3-1.3c0-0.7,0.6-1.3,1.3-1.3h5.8c0.7,0,1.3,0.6,1.3,1.3
            C46.4,69.2,45.8,69.8,45.1,69.8z"
          />
        </svg>
      </div>
    </div>
  </div>
  <div
    class="side-control"
  >
    <div
      class="bp3-form-group"
    >
      <div
        class="bp3-form-content"
      >
        <div>
          <label
            class="bp3-control bp3-radio"
          >
            <input
              checked=""
              name="Blueprint3.RadioGroup-0"
              type="radio"
              value="countData"
            />
            <span
              class="bp3-control-indicator"
            />
            Segment count
          </label>
          <label
            class="bp3-control bp3-radio"
          >
            <input
              name="Blueprint3.RadioGroup-0"
              type="radio"
              value="sizeData"
            />
            <span
              class="bp3-control-indicator"
            />
            Total size
          </label>
        </div>
      </div>
    </div>
    <div
      class="bp3-form-group"
    >
      <label
        class="bp3-label"
      >
        Datasource:
         
        <span
          class="bp3-text-muted"
        />
      </label>
      <div
        class="bp3-form-content"
      >
        <div
          class="bp3-html-select bp3-fill"
        >
          <select>
            <option
              value="all"
            >
              Show all
            </option>
          </select>
          <span
            class="bp3-icon bp3-icon-double-caret-vertical"
            icon="double-caret-vertical"
          >
            <svg
              data-icon="double-caret-vertical"
              height="16"
              viewBox="0 0 16 16"
              width="16"
            >
              <desc>
                double-caret-vertical
              </desc>
              <path
                d="M5 7h6a1.003 1.003 0 00.71-1.71l-3-3C8.53 2.11 8.28 2 8 2s-.53.11-.71.29l-3 3A1.003 1.003 0 005 7zm6 2H5a1.003 1.003 0 00-.71 1.71l3 3c.18.18.43.29.71.29s.53-.11.71-.29l3-3A1.003 1.003 0 0011 9z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </div>
      </div>
    </div>
    <div
      class="bp3-form-group"
    >
      <label
        class="bp3-label"
      >
        Period:
         
        <span
          class="bp3-text-muted"
        />
      </label>
      <div
        class="bp3-form-content"
      >
        <div
          class="bp3-html-select bp3-fill"
        >
          <select>
            <option
              value="1"
            >
               1 months
            </option>
            <option
              value="3"
            >
               3 months
            </option>
            <option
              value="6"
            >
               6 months
            </option>
            <option
              value="9"
            >
               9 months
            </option>
            <option
              value="12"
            >
               1 year
            </option>
          </select>
          <span
            class="bp3-icon bp3-icon-double-caret-vertical"
            icon="double-caret-vertical"
          >
            <svg
              data-icon="double-caret-vertical"
              height="16"
              viewBox="0 0 16 16"
              width="16"
            >
              <desc>
                double-caret-vertical
              </desc>
              <path
                d="M5 7h6a1.003 1.003 0 00.71-1.71l-3-3C8.53 2.11 8.28 2 8 2s-.53.11-.71.29l-3 3A1.003 1.003 0 005 7zm6 2H5a1.003 1.003 0 00-.71 1.71l3 3c.18.18.43.29.71.29s.53-.11.71-.29l3-3A1.003 1.003 0 0011 9z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
